AnkiDroid/src/main/java/com/ichi2/anki/CrashReportService.kt

```diff
@@ -257,34 +265,93 @@ object CrashReportService {
         val reportMode = mApplication.applicationContext.sharedPrefs()
             .getString(FEEDBACK_REPORT_KEY, FEEDBACK_REPORT_ASK)
+
         if (onlyIfSilent) {
             if (FEEDBACK_REPORT_ALWAYS != reportMode) {
                 Timber.i("sendExceptionReport - onlyIfSilent true, but ACRA is not 'always accept'. Skipping report send.")
-                return
+                return false
             }
         }
+
         if (FEEDBACK_REPORT_NEVER != reportMode) {
+            if (processException(e) == null) {
+                return false
+            }
+
             ACRA.errorReporter.putCustomData("origin", origin ?: "")
             ACRA.errorReporter.putCustomData("additionalInfo", additionalInfo ?: "")
             ACRA.errorReporter.handleException(e)
+            return true
         }
+
+        return false
+    }
+
+    fun Throwable.replaceMessage(newMessage: String?): Throwable =
+        Throwable(newMessage, this.cause)
+
+    private val exceptionRules: List<(Throwable) -> Throwable?> = listOf(
+        { e ->
+            // BackendSyncServerMessage may contain PII and we do not want this leaked to ACRA.
+            // If the exception is of a cause deemed to be associated with a message containing PII,
+            // then do not send it at all to retain user privacy.
+            // Related: https://github.com/ankidroid/Anki-Android/issues/17392
+            // and also https://github.com/ankitects/anki/commit/ba1f5f4
+            if (e.cause is BackendSyncServerMessageException) null else e
+        },
+        { e ->
+
+            // Make sure we filter out any email addresses which may be leaked by the sync server.
+            e.replaceMessage(
+                Patterns.EMAIL_ADDRESS.toRegex().replace(e.message.orEmpty(), "-EMAIL-")
+            )
+        },
+        { e ->
+            // Filter out the current username in the message (if it exists) as it is considered PII.
+            e.replaceMessage(
+                e.message.orEmpty()
+                    .replace(mPreferences?.get(SyncPreferences.USERNAME).toString(), "-USERNAME-")
+            )
+        }
+    )
+
+    // Apply each and every rule in exceptionRules on the original exception.
+    // This means that after this method call, we can get an exception that has the username and
+    // email of the user filtered out so it doesn't get sent to ACRA as is.
+    // If the method returns null, that means we do not send a report to ACRA at all as we have
+    // decided to block this based on some arbitrary rule(s) enforced on the exception's contents.
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    fun processException(e: Throwable): Throwable? {
+        var newE: Throwable? = e
+
+        for (rule in exceptionRules) {
+            newE = rule.invoke(newE!!)
+
+            if (newE == null) {
+                return null
+            }
+        }
+
+        return newE
```

Comment:
Looks like this started with Exception but is all Throwable now so should be altered (plus callers) to reflect
Also the comment strongly binds the implementation with the idea of the rules so this comment will need an update as rules change, or (better) comment should be more generic

```suggestion
    // Strip PII from Throwable.message or return null if we should discard completely
    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
    fun processThrowable(t: Throwable): Throwable? {
        var newT: Throwable? = t

        for (rule in throwableRules) {
            newT = rule.invoke(newT!!)

            if (newT == null) {
                return null
            }
        }

        return newT
```

Reply:
No clue what this patch is supposed to reflect, except "use the term Throwable more than Exception". I can do that.

Reply:
The patch was intended to reflect what I stated:

> Looks like this started with Exception but is all Throwable now so should be altered (plus callers) to reflect
Also the comment strongly binds the implementation with the idea of the rules so this comment will need an update as rules change, or (better) comment should be more generic

That is:
- the comment of this function shouldn't be bound to the implementation of another function
- it's all about Throwable in implementation but uses "Exception" in words

I believe those important but also think they're close to bikeshedding and when I bikeshed something I try to post a concrete suggestion that can just be adopted since (in my world at least) if you're going to bikeshed the onus is on you (that is: me in this context) to reduce back-and-forth and PR submitter effort trying to satisfy some bikeshed. So I made a whole suggestion.

Appears the comment part was ignored and a lot of the 'e's are still in there when it's really a 't', so that's a missed opportunity as the suggestion just did it all in one swell foop